enzyme/Enzyme/MLIR/Dialect/Ops.cpp

```diff
@@ -187,6 +175,118 @@ static inline bool isMutable(Type type) {
   return false;
 }
 
+/**
+ *
+ * Modifies input activites for the FwdDiffOp
+ * The activity promotion flow is as follows
+ * (depending on variable use):
+ *
+ *           -----> enzyme_dupnoneed
+ *          /              /
+ * enzyme_dup            /
+ *          \          v
+ *           ------> enzyme_const
+ *
+ */
+class FwdInpOpt final : public OpRewritePattern<ForwardDiffOp> {
+public:
+  using OpRewritePattern<ForwardDiffOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(ForwardDiffOp uop,
+                                PatternRewriter &rewriter) const override {
+
+    if (uop.getOutputs().size() == 0)
+      return failure();
+
+    auto retActivity = uop.getRetActivity();
+    auto inActivity = uop.getActivity();
+
+    auto in_idx = 0;
+    SmallVector<mlir::Value, 2> in_args;
+    SmallVector<ActivityAttr, 2> newInActivityArgs;
+    bool changed = false;
+    for (auto [idx, act] : llvm::enumerate(inActivity)) {
+      auto iattr = cast<ActivityAttr>(act);
+      auto val = iattr.getValue();
+
+      // Forward mode Input activities can only take values {dup, dupnoneed,
+      // const }
+
+      mlir::Value inp = uop.getInputs()[in_idx];
+
+      switch (val) {
+
+      case mlir::enzyme::Activity::enzyme_const:
+        in_args.push_back(inp);
+        newInActivityArgs.push_back(iattr);
+        break;
+
+      case Activity::enzyme_dupnoneed: {
+        // always pass in primal
+        in_args.push_back(inp);
+        in_idx++;
+
+        // selectively push or skip directional derivative
+        inp = uop.getInputs()[in_idx];
+        if (!isMutable(inp.getType()) &&
+            (matchPattern(inp, m_Zero()) ||
```

Comment:
@avik-pal @Pangoraw do you remember the is zero we had set up [since this won't applies for complex as is]